### PR TITLE
Add toolchain support for clang-cl on Windows

### PR DESCRIPTION
**Pull request by meteorcloudy**
Previously we [added clang-cl.exe support](https://github.com/bazelbuild/bazel/pull/6553) through `USE_CLANG_CL` environment variable. That is a "hack" on the existing MSVC toolchain to make it use Clang tools. This change add proper toolchain support for `clang-cl.exe` on Windows.
Because `clang-cl.exe` is MSVC compatible, we can just reuse the MSVC toolchain definition.

After this change, to select the clang-cl toolchain on Windows, you can use:

-  Before Bazel 1.0, specify build flag `--compiler=clang-cl`, this is deprecated and will be disabled in Bazel 1.0. 
-  From Bazel 1.0 (or with `--incompatible_enable_cc_toolchain_resolution` flag). You have to add a platform target to your build file (eg. the top level BUILD file):
```
platform(
  name = "windows-clang-cl",
  constraint_values = [
    "@platforms//cpu:x86_64",
    "@platforms//os:windows",
    "@bazel_tools//tools/cpp:clang-cl",
  ]
)
```
Then you can tell Bazel to use the clang-cl toolchain by specifying build flags
```
--extra_toolchains=@local_config_cc//:cc-toolchain-x64_windows-clang-cl --extra_execution_platforms=//:windows-clang-cl
```
or registering the platform and toolchain in your WORKSPACE file:
```
register_execution_platforms(
  ":windows-clang-cl"
)

register_toolchains(
    "@local_config_cc//:cc-toolchain-x64_windows-clang-cl",
)
```
Related issue: [incompatible_enable_cc_toolchain_resolution: Turn on toolchain resolution for cc rules](https://github.com/bazelbuild/bazel/issues/7260)

### CODE REVIEW

File: tools/cpp/windows_cc_configure.bzl [hunk 1]
```diff
@@ -575,6 +584,77 @@ def _get_msvc_vars(repository_ctx, paths):
...
+    clang_dir = llvm_path + "\\lib\\clang\\" + clang_version
+    clang_include_path = (clang_dir + "\\include").replace("\\", "\\\\")
+    clang_lib_path = (clang_dir + "\\lib\\windows").replace("\\", "\\\\")
+
+    clang_cl_vars = {
+        "%{clang_cl_env_tmp}": msvc_vars["%{msvc_env_tmp}"],
+        "%{clang_cl_env_path}": msvc_vars["%{msvc_env_path}"],
+        "%{clang_cl_env_include}": msvc_vars["%{msvc_env_include}"] + ";" + clang_include_path,
+        "%{clang_cl_env_lib}": msvc_vars["%{msvc_env_lib}"] + ";" + clang_lib_path,
+        "%{clang_cl_cxx_builtin_include_directories}": msvc_vars["%{msvc_cxx_builtin_include_directories}"] + (",\n        \"%s\"" % clang_include_path),
```
Comment (laszlocsomor) [inline tools/cpp/windows_cc_configure.bzl]: Do you need spaces, like in
https://github.com/bazelbuild/bazel/blob/594c32d8652b4bc5a8ad8f0566cf48c4458c40ee/tools/cpp/windows_cc_configure.bzl#L574
?

Comment (meteorcloudy) [reply tools/cpp/windows_cc_configure.bzl]: The leading spaces are already added in `msvc_vars["%{msvc_cxx_builtin_include_directories}"]` so we don't need them again.

File: tools/cpp/BUILD.windows.tpl [hunk 2]
```diff
@@ -179,6 +180,7 @@ cc_toolchain_config(
         "objdump": "wrapper/bin/msvc_nop.bat",
         "strip": "wrapper/bin/msvc_nop.bat",
     },
+    default_link_flags = ["/MACHINE:X64"],
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply tools/cpp/BUILD.windows.tpl]: This is not a bug fix, I added this new attribute `default_link_flags`, because `clang-cl` requires `/DEFAULTLIB:clang_rt.builtins-x86_64.lib` as well. And for later toolchain for arm and x86, this would also change to `/MACHINE:x86` and `/MACHINE:ARM`

File: tools/cpp/clang_installation_error.bat.tpl [hunk 3]
```diff
@@ -0,0 +1,24 @@
+:: Copyright 2017 The Bazel Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply tools/cpp/clang_installation_error.bat.tpl]: Done

**Comment by meteorcloudy** [comment]
I'll also add some documentation for this. Will ping the thread when it's done!

**Comment by meteorcloudy** [comment]
Ready for review again, tests will probably fail on Windows until we have the new Windows image with https://github.com/bazelbuild/continuous-integration/commit/baac80c3ee67cf27c3cc778e55c0547b7bfb4ebe

File: site/docs/windows.md [hunk 4]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
     set BAZEL_WINSDK_FULL_VERSION=10.0.10240.0
     ```
 
+*   The Clang compiler
+    
+    From 0.29.0, Bazel supports building with LLVM's MSVC-compatible compiler driver (`clang-cl.exe`).
+    To build with `clang-cl.exe`, you have to install **both** LLVM and Visual C++ Build tools.
```
Comment (laszlocsomor) [inline site/docs/windows.md]: Could you rephrase it as a `**Requirement**:` section, like above?

File: site/docs/windows.md [hunk 5]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
...
+    Because we still need to link to Visual C++ libraries.
```
Comment (laszlocsomor) [inline site/docs/windows.md]: I think this should belong to the previous sentence.

File: site/docs/windows.md [hunk 6]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
...
+
+    To install LLVM, please visit http://releases.llvm.org/download.html.
+    
+    Bazel can automatically detect LLVM installation on your system, you can explicitly tell Bazel where LLVM is installed by `BAZEL_LLVM`.
```
Comment (laszlocsomor) [inline site/docs/windows.md]: Missing word? Did you mean:

Bazel can automatically detect LLVM installation on your system, **[or]** you can explicitly tell Bazel (...)

File: site/docs/windows.md [hunk 7]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
...
+        ```
+    
+    To use the Clang toolchain for building C++, there are two situations.
+    
+    * In Bazel 0.29.0: You can enable the Clang toolchain by a build flag `--compiler=clang-cl`. 
+      This is deprecated and will be removed in Bazel 1.0.
+    
+    * From Bazel 1.0: You have to add a platform target to your build file (eg. the top level BUILD file):
+        ```
+        platform(
```
Comment (laszlocsomor) [inline site/docs/windows.md]: please format with Buildifier

File: site/docs/windows.md [hunk 8]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
...
+    
+    *   `BAZEL_LLVM` the LLVM installation directory
+
+        ```
+        set BAZEL_LLVM=C:\Program Files\LLVM
+        ```
+    
+    To use the Clang toolchain for building C++, there are two situations.
+    
+    * In Bazel 0.29.0: You can enable the Clang toolchain by a build flag `--compiler=clang-cl`.
```
Comment (laszlocsomor) [inline site/docs/windows.md]: Could you add a section `* In bazel 0.28 and older: Clang is not supported`?

File: site/docs/windows.md [hunk 9]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
...
+            "@platforms//os:windows",
+            "@bazel_tools//tools/cpp:clang-cl",
+          ]
+        )
+        ```
+        Then you can enable the Clang toolchain by specifying the following build flags:
+        ```
+        --extra_toolchains=@local_config_cc//:cc-toolchain-x64_windows-clang-cl --extra_execution_platforms=//:windows-clang-cl
+        ```
+        **or** registering the platform and toolchain in your WORKSPACE file:
```
Comment (laszlocsomor) [inline site/docs/windows.md]: what flags are needed in this case? Please clarify in the doc

File: src/test/py/bazel/bazel_windows_cpp_test.py [hunk 10]
```diff
@@ -592,6 +592,56 @@ def testCppErrorShouldBeVisible(self):
     self.AssertExitCode(exit_code, 1, stderr)
     self.assertIn('this_is_an_error', ''.join(stdout))
 
+  def testBuildWithClangClByCompilerFlag(self):
```
Comment (laszlocsomor) [inline src/test/py/bazel/bazel_windows_cpp_test.py]: Could you either disable these tests until the CI machines have Clang, or wait with merging this PR until then?

If you merge now, everyone's presubmits will fail on Windows because of these tests.

File: site/docs/windows.md [hunk 11]
```diff
@@ -198,6 +198,54 @@ To build C++ targets, you need:
     set BAZEL_WINSDK_FULL_VERSION=10.0.10240.0
     ```
 
+*   The Clang compiler
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply site/docs/windows.md]: Make sense, I added `Build C++ with Clang` section, and renamed original `Build C++` to `Build C++ with MSVC`

File: site/docs/windows.md (hunk 4 above)
Comment (meteorcloudy) [reply site/docs/windows.md]: Done

File: site/docs/windows.md (hunk 6 above)
Comment (meteorcloudy) [reply site/docs/windows.md]: Done

File: site/docs/windows.md (hunk 8 above)
Comment (meteorcloudy) [reply site/docs/windows.md]: Done

File: site/docs/windows.md (hunk 9 above)
Comment (meteorcloudy) [reply site/docs/windows.md]: In Bazel 1.0, you don't need any other flag, because `--incompatible_enable_cc_toolchain_resolution` will be enabled by default. I reformat the doc to make this more clear.

File: src/test/py/bazel/bazel_windows_cpp_test.py (hunk 10 above)
Comment (meteorcloudy) [reply src/test/py/bazel/bazel_windows_cpp_test.py]: Yes, I'll wait until the CI machines have Clang.

File: site/docs/windows.md (hunk 7 above)
Comment (meteorcloudy) [reply site/docs/windows.md]: Done

File: site/docs/windows.md (hunk 5 above)
Comment (meteorcloudy) [reply site/docs/windows.md]: Done
